Sources/RInAppMessaging/Views/TooltipView.swift

```diff
@@ -0,0 +1,320 @@
+            path.addLine(to: CGPoint(x: 0, y: bounds.midY))
+            path.addLine(to: CGPoint(x: UIConstants.tipSize.height, y: bounds.midY + UIConstants.tipSize.width / 2.0))
+        }
+
+        path.fill()
+        layer.shadowPath = path.cgPath
+    }
+
+    override func didMoveToSuperview() {
+        super.didMoveToSuperview()
+        guard let superview = superview else {
+            return
+        }
+
+        exitButton.removeFromSuperview()
+        superview.addSubview(exitButton)
+        var constraints = [
+            exitButton.widthAnchor.constraint(equalToConstant: UIConstants.exitButtonSize),
+            exitButton.heightAnchor.constraint(equalToConstant: UIConstants.exitButtonSize)
+        ]
+
+        if [.left, .topLeft, .bottomLeft].contains(position) {
+            constraints.append(exitButton.rightAnchor.constraint(equalTo: self.leftAnchor,
+                                                                 constant: -UIConstants.exitButtonRightMargin))
+        } else {
+            constraints.append(exitButton.leftAnchor.constraint(equalTo: self.rightAnchor,
+                                                                constant: UIConstants.exitButtonRightMargin))
+        }
+        if [.bottomRight, .bottomLeft].contains(position) {
+            constraints.append(exitButton.bottomAnchor.constraint(equalTo: self.topAnchor,
+                                                                  constant: -UIConstants.exitButtonTopMargin + UIConstants.tipSize.height))
+        }
+        else if position == .bottomCentre {
+            constraints.append(exitButton.bottomAnchor.constraint(equalTo: self.bottomAnchor,
+                                                                  constant: -UIConstants.exitButtonTopMargin + UIConstants.tipSize.height))
+
+        } else {
+            constraints.append(exitButton.bottomAnchor.constraint(equalTo: self.topAnchor,
+                                                                  constant: -UIConstants.exitButtonTopMargin))
+        }
+        NSLayoutConstraint.activate(constraints)
+    }
+
+    override func removeFromSuperview() {
+        super.removeFromSuperview()
+        exitButton.removeFromSuperview()
+    }
+
+    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+        if exitButton.isTouchInside(touchPoint: point, from: self, touchAreaSize: UIConstants.exitButtonTouchAreaSize) == true {
+            return exitButton
+        }
+        return super.hitTest(point, with: event)
+    }
+
+    // MARK: - Interface methods
+
+    func startAutoFadingIfNeeded(seconds: UInt) {
+        guard autoFadeTimer == nil else {
+            return
+        }
+
+        autoFadeTimer = Timer(fire: Date().addingTimeInterval(TimeInterval(seconds)), interval: 0, repeats: false, block: { [weak self] _ in
+            self?.onExitButtonTap()
+        })
+
+        RunLoop.current.add(autoFadeTimer!, forMode: .common)
```

Comment:
Shouldn't force unwrap here - can we catch these with SwiftLint?

Reply:
oh - the Timer init can't fail. Maybe it's OK then

Reply:
safe but I refactored it so the force unwrap is avioded